Improve development workflow

cassava opened the pull request:
In particular with a focus on debugging.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/eclipse/cloe/31)
<!-- Reviewable:end -->

cassava reviewed (commented):
If we could get this merged today or tomorrow that would be good. Then I can tag a 0.18.0-rc1 pre-release so that we aren't working with a moving target all the time.

> *[Reviewable](https://reviewable.io/reviews/eclipse/cloe/31#-:-M_0eTgv1k9cflf_MIW7:b-ia5c4q)* status: 0 of 16 files reviewed, all discussions resolved (waiting on @clssn and @tobifalk)



<!-- Sent from Reviewable.io -->

tobifalk commented on dist/docker/Dockerfile.focal:
```diff
@@ -0,0 +1,89 @@
+# Dockerfile.ubuntu
+#
+# This file acts partly as a Docker recipe for building Cloe on Ubuntu.
+# It is currently only tested with Ubuntu 18.04 and 20.04.
```
This comment can probably be removed.

cassava replied:
Done.

tobifalk commented on dist/docker/Dockerfile.alpine:
```diff
@@ -45,11 +45,14 @@ RUN pip3 install --upgrade pip && \
 WORKDIR /cloe
 COPY . /cloe
+ARG WITH_VTD=0
+ARG PACKAGE_TARGET=package-select
```
I think this is not going to work, s. our previous discussion [here](https://github.com/eclipse/cloe/pull/24#discussion_r582051715). There will be conflicting boost versions used for oak and cloe-runtime.

cassava replied:
Right you are. I created a new target, `package-auto`, that builds all selected packages together.

tobifalk commented on vendor/vtd/conanfile.py:
```diff
@@ -43,17 +46,48 @@ def configure(self):
+        binary_files = [Path(x) for x in result.stdout.decode().splitlines()]
+        # Set RPATH of all the collected binaries:
+        for file in binary_files:
+            assert file.exists()
```
Here one could add 
```
            if not ".so" in file.suffixes:
                continue
```
to avoid a bunch of error messages when it tries to set RPATH e.g. on pdf files.

tobifalk commented on vendor/vtd/conanfile.py:
```diff
@@ -43,17 +46,48 @@ def configure(self):
+            libraries = [Path(x.strip()) for x in file.read().splitlines()]
+        for path in libraries:
+            print(f"Bundling system library: {path}")
+            shutil.copy(path, libdir, follow_symlinks=False)
```
This does not work on my machine since `library_filelist.txt` contains a bunch of libraries that I don't have or where I have a different version.

cassava replied:
This is a problem; one way around this is to only specify the symlinks to copy and resolve these into concrete paths. I could do this.

cassava replied:
I implemented this now, so it should work a little better.

cassava reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/eclipse/cloe/31)* status: 0 of 22 files reviewed, 2 unresolved discussions (waiting on @clssn and @tobifalk)

___
*[vendor/vtd/conanfile.py, line 73 at r4](https://reviewable.io/reviews/eclipse/cloe/31#-M_uEVvpYN-58gBJptDK-r4-73:-M_uqqWWD2OAsR0iI_1w:b-896fix) ([raw file](https://github.com/eclipse/cloe/blob/23ab0641c069f8654642b48ece927d8d6844853a/vendor/vtd/conanfile.py#L73)):*
<details><summary><i>Previously, cassava (Ben Morgan) wrote…</i></summary><blockquote>

Good idea, thanks! LGTM.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

cassava commented on Makefile.package:
```diff
@@ -105,10 +108,32 @@ help:
+status:
+	# Show the status of each package in the cloe workspace:
+	# - Is it editable / built / packaged?
+	# - Is it available in debug mode?
```
I cleaned up the comment.

cassava commented on Makefile.package:
```diff
@@ -105,10 +108,32 @@ help:
+	#
+	_editable=$$(conan editable list | grep "${PACKAGE_FQN}"); \
+	if [ -z "$${_editable}" ]; then \
+		if [ -n "${PACKAGE_DATE}" ] && [ -z "$$(find -type f -newermt "${PACKAGE_DATE}")" ]; then \
```
You have some very valuable points here.
- I didn't want to do an export because that would modify the cache, and I don't want to do that for an operation that should be read-only.
- I want to let the user know that the package in the cache is *different* to the current source, so if I use the git log it will show the package as up-to-date if I ever switch to a branch that is older than the one that created the last build.
- Of course this also means if you checkout A, build, checkout B, checkout A -> make status will show in the worst case that everything as outdated

I'm going to leave this as is, but feel free to submit a PR improving it.

cassava commented on Makefile.package:
```diff
@@ -41,9 +41,11 @@ BUILD_LAYOUT     := ${PROJECT_ROOT}/.conan-layout.ini
 #
 # This can be one of: None, Debug, Release, RelWithDebInfo, MinSizeRel
+BUILD_TYPE := RelWithDebInfo
 DEBUG_BUILD_TYPE := Debug
```
I'll remove it.

cassava commented on Makefile.package:
```diff
@@ -41,9 +41,11 @@ BUILD_LAYOUT     := ${PROJECT_ROOT}/.conan-layout.ini
+# only apply to local builds.
 #
 # This can be one of: None, Debug, Release, RelWithDebInfo, MinSizeRel
+BUILD_TYPE := RelWithDebInfo
```
I don't know, this is currently orthogonal with the other local build variables, like `BUILD_DIR`

cassava replied:
Leaving it for now.

cassava commented on cli/cloe_launch/__main__.py:
```diff
@@ -114,11 +114,25 @@ def deny_profile_and_path(profile: str, profile_path: str) -> None:
     help="Options to pass to Conan for virtualrunenv generation",
 )
+@click.option(
+    "-o:s",
```
Yes, exactly. Of course, this assumes that a package `virtue:build_type=Debug` has already been built.

cassava commented:
I have addressed the points from Tobias and tested the state across the Docker images; unfortunately Tobias is currently not able to update his review on Github at the moment, so I have received confirmation through another channel that I can merge the PR.
